openvm/src/customize_exe.rs

```diff
@@ -584,3 +612,164 @@ fn transpose_symbolic_machine<F: PrimeField32, P: FieldElement>(
         bus_interactions,
     }
 }
+
+fn sort_blocks_by_pgo_cell_cost<F: PrimeField32>(
+    blocks: &mut Vec<BasicBlock<F>>,
+    pgo_program_idx_count: HashMap<u32, u32>,
+    airs: &BTreeMap<usize, SymbolicMachine<powdr_number::BabyBearField>>,
+    config: PowdrConfig,
+    opcodes_no_apc: &[usize],
+    branch_opcodes_set: &BTreeSet<usize>,
+) -> HashMap<usize, CachedAutoPrecompile<powdr_number::BabyBearField>> {
+    // drop any block whose start index cannot be found in pc_idx_count,
+    // because a basic block might not be executed at all.
+    // Also only keep basic blocks with more than one original instruction.
+    blocks.retain(|b| {
+        pgo_program_idx_count.contains_key(&(b.start_idx as u32)) && b.statements.len() > 1
+    });
+
+    tracing::info!(
+        "Retained {} basic blocks after filtering by pc_idx_count",
+        blocks.len()
+    );
+
+    // store air width by opcode, so that we don't repetitively calculate them later
+    // filter out opcodes that contain next references in their air, because they are not supported yet in apc
+    let air_width_by_opcode = airs
+        .iter()
+        .filter(|&(i, _)| (!opcodes_no_apc.contains(i)))
+        .map(|(i, air)| (*i, air.unique_columns().count()))
+        .collect::<HashMap<_, _>>();
+
+    // generate apc and cache it for all basic blocks
+    // calculate number of trace cells saved per row for each basic block to sort them by descending cost
+    let (cells_saved_per_row_by_bb, apcs): (HashMap<_, _>, HashMap<_, (_, _, _)>) = blocks
+        .iter()
+        .enumerate()
+        .filter_map(|(i, acc_block)| {
```

Comment:
This block is quite large now and is doing many different things, I think it'd be good to split it up